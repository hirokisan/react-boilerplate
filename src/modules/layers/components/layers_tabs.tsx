import React, { FC } from 'react';

import { classes, styled, Action, numberAttribute, theme } from 'config/common';

const LayerHeader = styled.div`
  width: 100%;
  height: 40px;
  min-height: 40px;
  display: flex;
  align-items: center;
  justify-content: center;
`;

const LayerList = styled.ul`
  width: 100%;
  height: 100%;
  align-items: center;
  display: flex;
  justify-content: flex-start;
`;

const LayerItem = styled.li`
  height: 100%;
  align-items: center;
  color: rgba(255, 255, 255, 0.4);
  border-left: 1px solid ${theme.paneBackground};
  border-right: 1px solid ${theme.divider};
  cursor: pointer;
  display: flex;
  font-size: 11px;
  font-weight: 400;
  justify-content: center;
  line-height: 15px;
  padding: 0 12px;
  transition: background-color 0.2s ease-out;

  &:hover,
  &.active {
    background-color: rgba(77, 77, 77, 0.17);
    border-left-color: ${theme.divider};
  }
  &.active {
    border-right-color: transparent;
    border-left: 1px solid rgba(77, 77, 77, 0.17);
    color: ${theme.paneText};
  }
`;

type Props = {
  /**
   * Id of the current tab
   *
   * @default 0
   */
  currentTab?: number;
  /**
   * Callback for setting the active tab
   */
  setTab: Action<number>;
  /**
   * List of tabs to display
   */
  tabs: string[];
};

const defaultTabs = ['LAYERS', 'INDICATORS'];

export const LayersTabs: FC<Props> = ({ currentTab = 0, setTab, tabs = defaultTabs }) => {
  const setActiveTab = React.useCallback(
    (e: React.MouseEvent<HTMLLIElement>) => {
      setTab(numberAttribute(e.currentTarget, 'data-tab'));
    },
    [setTab]
  );

  return (
    <>
      <LayerHeader>
        <LayerList>
          {tabs.map((tab, i) => (
            <LayerItem
              key={i}
              className={classes({ active: currentTab === i })}
              data-tab={i}
              onClick={setActiveTab}
            >
              {tab}
            </LayerItem>
          ))}
        </LayerList>
      </LayerHeader>
    </>
  );
};
