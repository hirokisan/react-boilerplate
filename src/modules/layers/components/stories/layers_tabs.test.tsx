import React from 'react';
import renderer from 'react-test-renderer';

import { LayersTabs } from '../layers_tabs';

describe('Layer Tabs', () => {
  it('renders default', () => {
    const tree = renderer.create(
      <LayersTabs currentTab={0} setTab={jest.fn()} tabs={['LAYERS', 'INDICATORS']} />
    );
    expect(tree).toMatchSnapshot();
  });

  it('changes active tab by index', () => {
    const tree = renderer.create(
      <LayersTabs currentTab={1} setTab={jest.fn()} tabs={['LAYERS', 'INDICATORS']} />
    );
    expect(tree.root.findByProps({ 'data-tab': 1 }).props.className).toEqual('active');
    expect(tree).toMatchSnapshot();
  });

  it('calls the change tab function', () => {
    const setTab = jest.fn();
    const tree = renderer.create(
      <LayersTabs currentTab={1} setTab={setTab} tabs={['LAYERS', 'INDICATORS']} />
    );

    const tabId = 3;
    const fakeEvent = {
      currentTarget: {
        hasAttribute: () => true,
        getAttribute: () => tabId
      }
    };
    tree.root.findByProps({ 'data-tab': 1 }).props.onClick(fakeEvent);

    expect(setTab).toHaveBeenCalledWith(tabId);
  });
});
