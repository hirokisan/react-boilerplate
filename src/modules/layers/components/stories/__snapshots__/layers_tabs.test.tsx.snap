// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layer Tabs changes active tab by index 1`] = `
<div
  className="sc-bdVaJa bFvmKt"
>
  <ul
    className="sc-bwzfXH jFumDk"
  >
    <li
      className="sc-htpNat AUdfI"
      data-tab={0}
      onClick={[Function]}
    >
      LAYERS
    </li>
    <li
      className="active sc-htpNat AUdfI"
      data-tab={1}
      onClick={[Function]}
    >
      INDICATORS
    </li>
  </ul>
</div>
`;

exports[`Layer Tabs renders default 1`] = `
<div
  className="sc-bdVaJa bFvmKt"
>
  <ul
    className="sc-bwzfXH jFumDk"
  >
    <li
      className="active sc-htpNat AUdfI"
      data-tab={0}
      onClick={[Function]}
    >
      LAYERS
    </li>
    <li
      className="sc-htpNat AUdfI"
      data-tab={1}
      onClick={[Function]}
    >
      INDICATORS
    </li>
  </ul>
</div>
`;
